planetiler-core/src/test/java/com/onthegomap/planetiler/pmtiles/PmtilesTest.java

```diff
@@ -96,44 +110,168 @@ void testRoundtripHeader() {
+    assertThrows(FileFormatException.class, () -> WriteablePmtiles.Header.fromBytes(new byte[127]));
   }
 
   @Test
   void testRoundtripDirectoryMinimal() {
-    ArrayList<Pmtiles.Entry> in = new ArrayList<>();
-    in.add(new Pmtiles.Entry(0, 0, 1, 1));
+    ArrayList<WriteablePmtiles.Entry> in = new ArrayList<>();
+    in.add(new WriteablePmtiles.Entry(0, 0, 1, 1));
 
-    List<Pmtiles.Entry> out = Pmtiles.deserializeDirectory(Pmtiles.serializeDirectory(in));
+    List<WriteablePmtiles.Entry> out = WriteablePmtiles.deserializeDirectory(WriteablePmtiles.serializeDirectory(in));
     assertEquals(in, out);
   }
 
   @Test
   void testRoundtripDirectorySimple() {
-    ArrayList<Pmtiles.Entry> in = new ArrayList<>();
+    ArrayList<WriteablePmtiles.Entry> in = new ArrayList<>();
 
     // make sure there are cases of contiguous entries and non-contiguous entries.
-    in.add(new Pmtiles.Entry(0, 0, 1, 0));
-    in.add(new Pmtiles.Entry(1, 1, 1, 1));
-    in.add(new Pmtiles.Entry(2, 3, 1, 1));
+    in.add(new WriteablePmtiles.Entry(0, 0, 1, 0));
+    in.add(new WriteablePmtiles.Entry(1, 1, 1, 1));
+    in.add(new WriteablePmtiles.Entry(2, 3, 1, 1));
 
-    List<Pmtiles.Entry> out = Pmtiles.deserializeDirectory(Pmtiles.serializeDirectory(in));
+    List<WriteablePmtiles.Entry> out = WriteablePmtiles.deserializeDirectory(WriteablePmtiles.serializeDirectory(in));
     assertEquals(in, out);
-    out = Pmtiles.deserializeDirectory(Pmtiles.serializeDirectory(in, 0, in.size()));
+    out = WriteablePmtiles.deserializeDirectory(WriteablePmtiles.serializeDirectory(in, 0, in.size()));
     assertEquals(in, out);
   }
 
   @Test
   void testRoundtripDirectorySlice() {
-    ArrayList<Pmtiles.Entry> in = new ArrayList<>();
+    ArrayList<WriteablePmtiles.Entry> in = new ArrayList<>();
 
     // make sure there are cases of contiguous entries and non-contiguous entries.
-    in.add(new Pmtiles.Entry(0, 0, 1, 0));
-    in.add(new Pmtiles.Entry(1, 1, 1, 1));
-    in.add(new Pmtiles.Entry(2, 3, 1, 1));
+    in.add(new WriteablePmtiles.Entry(0, 0, 1, 0));
+    in.add(new WriteablePmtiles.Entry(1, 1, 1, 1));
+    in.add(new WriteablePmtiles.Entry(2, 3, 1, 1));
 
-    List<Pmtiles.Entry> out = Pmtiles.deserializeDirectory(Pmtiles.serializeDirectory(in, 1, 2));
+    List<WriteablePmtiles.Entry> out = WriteablePmtiles.deserializeDirectory(
+      WriteablePmtiles.serializeDirectory(in, 1, 2));
     assertEquals(1, out.size());
   }
+
+  @Test
+  void testBuildDirectoriesRootOnly() throws IOException {
+
+  }
+
+  @Test
+  void testBuildDirectoriesLeaves() throws IOException {
+
+  }
+
+  @Test
+  void testWritePmtilesSingleEntry() throws IOException {
+    var bytes = new SeekableInMemoryByteChannel(0);
+    var in = WriteablePmtiles.newWriteToMemory(bytes);
+
+    var config = PlanetilerConfig.defaults();
+    in.initialize(config, new TileArchiveMetadata(new Profile.NullProfile()), new LayerStats());
```

Comment:
This is verbose to set up a test, since PMTiles depends on an entire profile and metadata to write out the JSON metadata and full header; this isn't true of MBTiles since you can read/write to the `tiles` tables in isolation

Reply:
You could use the constructor that takes a name, description, attribution, version, strings, then make sure the values you read out of the readable archive match the ones you passed in?

Reply:
Yeah, trying to read the JsonMetadata back out for tests, but this throws a runtime exception:

```java
  public record JsonMetadata(
    @JsonProperty("vector_layers") List<LayerStats.VectorLayer> vectorLayers,
    @JsonAnyGetter @JsonAnySetter Map<String, String> otherMetadata
  ) {

    public byte[] toBytes() {

      try {
        return objectMapper.writeValueAsBytes(this);
      } catch (JsonProcessingException e) {
        throw new IllegalArgumentException("Unable to encode as string: " + this, e);
      }
    }

    public static JsonMetadata fromBytes(byte[] bytes) {
      try {
        return objectMapper.readValue(bytes, JsonMetadata.class);
      } catch (IOException e) {
        throw new IllegalStateException("Invalid metadata json: " + bytes, e);
      }
    }
  }
```

I'm letting Jackson serialize arbitrary string->string, and need to get it to behave the same way in the opposite direction

Reply:
Looks like this is because of https://github.com/FasterXML/jackson-databind/issues/3439 - there is a workaround described there where you just add an extra constructor with annotations to make it visible to jackson:

```java
    @JsonCreator
    JsonMetadata(@JsonProperty("vector_layers") List<LayerStats.VectorLayer> vectorLayers) {
      this(vectorLayers, new HashMap<>());
    }
```